server/market/market.go

```diff
@@ -3,30 +3,327 @@
+func NewMarket(ctx context.Context, epochDurationSec int64, baseAsset, quoteAsset *asset.Asset,
+	storage db.DEXArchivist, authMgr swap.AuthManager) (*Market, error) {
+
+	lotSize := baseAsset.LotSize
+	mktInfo, err := types.NewMarketInfoFromSymbols(baseAsset.Symbol, quoteAsset.Symbol, lotSize)
+	if err != nil {
+		return nil, err
+	}
+
+	swapperCfg := &swap.Config{
+		Ctx: ctx,
+		Assets: map[uint32]*asset.Asset{
+			mktInfo.Base:  baseAsset,
+			mktInfo.Quote: quoteAsset,
+		},
+		Storage:          storage,
+		AuthManager:      authMgr,
+		BroadcastTimeout: 10 * time.Second, // TODO: who sets this?
+	}
+
+	// Supplement the provided context with an cancel function. The Stop method
+	// may be used to stop the market, or the parent context itself may be
+	// canceled.
+	ctxInternal, cancel := context.WithCancel(ctx)
+
+	return &Market{
+		ctx:           ctxInternal,
+		cancel:        cancel,
+		marketInfo:    mktInfo,
+		lotSize:       lotSize,
+		epochDuration: epochDurationSec,
+		orderRouter:   make(chan *orderUpdateSignal, 16),
+		book:          book.New(lotSize),
+		matcher:       matcher.New(),
+		swapper:       swap.NewSwapper(swapperCfg),
+		auth:          authMgr,
+		baseAsset:     baseAsset,
+		quoteAsset:    quoteAsset,
+		storage:       storage,
+	}, nil
+}
+
+// Start beings order processing.
+func (m *Market) Start() {
+	m.wg.Add(1)
+	go m.runEpochs()
+}
+
+// Stop begins the Market's shutdown. Use WaitForShutdown after Stop to wait for
+// shutdown to complete.
+func (m *Market) Stop() {
+	log.Infof("Market shutting down...")
+	m.cancel()
+}
+
+// WaitForShutdown waits until the order processing is finished and the Market
+// is stopped.
+func (m *Market) WaitForShutdown() {
+	m.wg.Wait()
+}
+
+func (m *Market) notifyNewOrder(order order.Order, action bookUpdateAction) {
+	// send to each receiver
+	for _, s := range m.orderFeeds {
+		select {
+		case <-m.ctx.Done():
+			return
+		case s <- &bookUpdateSignal{action, order}:
+		}
+	}
+}
+
+// runEpochs is the main order processing loop. The queue, m.epochQueue, should
+// only be accessed in this loop.
+func (m *Market) runEpochs() {
+	// Startup epoch is the epoch after the epoch one second in the future.
+	dur := m.epochDuration
+	nextEpochIdx := 1 + (1+time.Now().Unix())/dur
+	nextEpoch := NewEpoch(nextEpochIdx, dur)
+	epochCycle := time.After(time.Until(nextEpoch.Start))
+
+	defer m.wg.Done()
+
+	for {
+		select {
+		case <-m.ctx.Done():
+			return
+
+		case <-epochCycle:
+			if m.epochQueue != nil {
+				go m.processEpoch(m.epochQueue)
+				// m.epochQueue must now be replaced as it has been handed off.
+			}
+			m.epochQueue = nextEpoch
+
+			nextEpochIdx = m.epochQueue.Epoch + 1
+			nextEpoch = NewEpoch(nextEpochIdx, dur)
+			epochCycle = time.After(time.Until(nextEpoch.Start))
+
+		case s := <-m.orderRouter:
+			// Set the server time stamp here, BEFORE referencing the order ID.
+			ord := s.rec.order
+			ord.SetTime(time.Now().Unix())
```

Comment:
Note that "now" may end up being slightly after when the epoch would close if the latency of the channel receive and the call to Now take some nano seconds too long.  I don't know how to guarantee this doesn't happen, but perhaps the order can be send back into m.orderRouter if the server time stamp computes to the nextEpochIdx instead of current.

Reply:
Resolved in following commits by checking the time and sending back on the channel to allow the epoch to cycle first.